HIP/hip-792.md

```diff
@@ -0,0 +1,290 @@
+---
+hip: 792
+title: Extend isAuthorized to process current transaction in HAS System Contract
+author: Brendan Graetz <@bguiz>
+working-group: Nana Essilfie-Conduah <@nana-ec>, Luke Lee  <@lukelee-sl>
+type: Standards Track
+category: Service
+needs-council-approval: Yes
+status: Draft
+last-call-date-time: 2023-10-22T16:00:00Z
+created: 2023-08-21
+discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/792
+updated: 2023-08-23
+requires: 632
+---
+
+## Abstract
+
+<!-- a short (~200 word) description of the technical issue being addressed. -->
+
+HIP-632 adds a new system contract, `hederaAccountService` that exposes a method `isAuthorized(address, messageHash, signatureBlob)`. This HIP proposes a similar version of that function, which implements the same functionality, but restricted for use *only* on the current (in flight) transaction. This exposes a new method `isAuthorizedCurrentTransaction()`.
+
+## Motivation
+
+<!-- The motivation is critical for HIPs that want to change the Hedera codebase or ecosystem. It should clearly explain why the existing specification is inadequate to address the problem that the HIP solves. HIP submissions without sufficient motivation may be rejected outright. -->
+
+`hederaAccountService.isAuthorized(address, messageHash, signatureBlob)` does work for the following scenarios:
+
+- When an account's admin key is "simple", and is comprised of either a single EdDSA key, or a single ECDSA key
+- When an account's admin key is "complex", and is comprised of multiple EdDSA keys and/or  ECDSA keys, combined using one or more `KeyList`s or `ThresholdKey`s.
+
+`hederaAccountService.isAuthorized(address, messageHash, signatureBlob)` does **not** work for the following scenarios:
+
+- When an account's admin key is "complex", and is comprised of multiple EdDSA keys and/or  ECDSA keys and/or **smart contract IDs**, combined using one or more `KeyList`s or `ThresholdKey`s.
+
+The proposed `hederaAccountService.isAuthorizedCurrentTransaction()` method aims to fulfil the above scenario, and specific use cases pertaining to this scenario will be elaborated upon in the [use cases section](#use-cases) below.
```

Comment:
not necessarily *last key slot*. consider these scenario where:

- the account has a `ThresholdKey([edDsaKey, ecDsaKey, scFooId, scBarId], 1)`
  - --> transaction is signed using either of the cryptographic keys, does not invoke either of the smart contracts --> ✅ 
  - --> transaction is signed using either of the cryptographic keys, does not invoke either of the smart contracts --> ✅ 
  - --> transaction is unsigned using either of the cryptographic keys, but invokes smart contract `scFooId` --> ✅ 
  - --> transaction is unsigned using either of the cryptographic keys, but invokes smart contract `scBarId` --> ✅ 
- the account has a `ThresholdKey([edDsaKey, ecDsaKey, scFooId, scBarId], 2)`
  - --> transaction is unsigned using either of the cryptographic keys, but invokes smart contract `scFooId` --> ❌ 
  - --> transaction is unsigned using either of the cryptographic keys, but invokes smart contract `scBarId` --> ❌  
  - --> transaction is signed using either of the cryptographic keys, but invokes smart contract `scFooId` --> ✅ 
  - --> transaction is signed using either of the cryptographic keys, but invokes smart contract `scBarId` --> ✅ 
- the account has a `ThresholdKey(ThresholdKey([edDsaKey, ecDsaKey], 1), ThresholdKey([scFooId, scBarId], 1), 2)`
  - --> transaction is signed using either of the cryptographic keys, and invokes either smart contract --> ✅ 
  - --> transaction is not signed using either of the cryptographic keys, and invokes either smart contract --> ❌  

> is the goal to check if the calling smart contract satisfies the last key slot of the calling EOA's complex key (keyList or threshold)?

Reply:
I think only your last scenario should ever return true. Since SCs don't actually sign and their authorization is based on the matching contract making the call there's no way to authorize if an operation needs more than 1 contractId to match and if all the necessary cryptographic keys haven't signed.

Reply:
Yes, precisely - I was enumerating different examples of when a smart contract ID may not be in the last key slot of an account's complex key, and where there may be more than one contract ID in an account's complex key.

> Since SCs don't actually sign and their authorization is based on the matching contract making the call there's no way to authorize if an operation needs more than 1 contractId to match and if all the necessary cryptographic keys haven't signed.